pkg/kubectl/cmd/diff.go

```diff
@@ -0,0 +1,460 @@
+	return nil
+}
+
+// Printer is used to print an object.
+type Printer struct{}
+
+// Print the object inside the writer w.
+func (p *Printer) Print(obj map[string]interface{}, w io.Writer) error {
+	if obj == nil {
+		return nil
+	}
+	data, err := yaml.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(data)
+	return err
+
+}
+
+// DiffVersion gets the proper version of objects, and aggregate them into a directory.
+type DiffVersion struct {
+	Dir  *Directory
+	Name string
+}
+
+// NewDiffVersion creates a new DiffVersion with the named version.
+func NewDiffVersion(name string) (*DiffVersion, error) {
+	dir, err := CreateDirectory(name)
+	if err != nil {
+		return nil, err
+	}
+	return &DiffVersion{
+		Dir:  dir,
+		Name: name,
+	}, nil
+}
+
+func (v *DiffVersion) getObject(obj Object) (map[string]interface{}, error) {
+	switch v.Name {
+	case "LIVE":
+		return obj.Live()
+	case "MERGED":
+		return obj.Merged()
+	case "LOCAL":
+		return obj.Local()
+	case "LAST":
+		return obj.Last()
+	}
+	return nil, fmt.Errorf("Unknown version: %v", v.Name)
+}
+
+// Print prints the object using the printer into a new file in the directory.
+func (v *DiffVersion) Print(obj Object, printer Printer) error {
+	vobj, err := v.getObject(obj)
+	if err != nil {
+		return err
+	}
+	f, err := v.Dir.NewFile(obj.Name())
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return printer.Print(vobj, f)
+}
+
+// Directory creates a new temp directory, and allows to easily create new files.
+type Directory struct {
+	Name string
+}
+
+// CreateDirectory does create the actual disk directory, and return a
+// new representation of it.
+func CreateDirectory(prefix string) (*Directory, error) {
+	name, err := ioutil.TempDir("", prefix+"-")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Directory{
+		Name: name,
+	}, nil
+}
+
+// NewFile creates a new file in the directory.
+func (d *Directory) NewFile(name string) (*os.File, error) {
+	return os.OpenFile(filepath.Join(d.Name, name), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0700)
+}
+
+// Delete removes the directory recursively.
+func (d *Directory) Delete() error {
+	return os.RemoveAll(d.Name)
+}
+
+// Object is an interface that let's you retrieve multiple version of
+// it.
+type Object interface {
+	Local() (map[string]interface{}, error)
+	Live() (map[string]interface{}, error)
+	Last() (map[string]interface{}, error)
+	Merged() (map[string]interface{}, error)
+
+	Name() string
+}
+
+// InfoObject is an implementation of the Object interface. It gets all
+// the information from the Info object.
+type InfoObject struct {
+	Info    *resource.Info
+	Encoder runtime.Encoder
+	Parser  *parse.Factory
+}
+
+var _ Object = &InfoObject{}
+
+func (obj InfoObject) toMap(data []byte) (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+	if len(data) == 0 {
+		return m, nil
+	}
+	err := json.Unmarshal(data, &m)
+	return m, err
+}
+
+func (obj InfoObject) Local() (map[string]interface{}, error) {
```

Comment:
@apelisse this assumption that VersionedObject is "local" seems questionable.  Can you explain why you're treating it this way?  Are you setting VersionedObject to represent something different than what most builders are assuming?

Reply:
Someone told me that might do what I want, but I have no idea what I'm doing. Feedback is welcome :-)

Reply:
@smarterclayton , @apatil , I dont know if was broken from the beginning or this proposed change broke it, but in 1.9.2 kubectl diff command shows nothing. Quick check finds that LOCAL and LIVE files written by differ are byte-to-byte the same, which is not surprising that currently `obj.Local()` and `obj.Live()` serialize very same `obj.Info.Object`.

Before https://github.com/kubernetes/kubernetes/pull/55660 (== in this original PR) `Local()` was serializing `obj.Info.VersionedObject` , which doesn't look very local to me either, but I have no idea about these internals.

Here is how I call diff during my tests:
```
kubectl -n kube-system apply -f  manifests/kube-dns-deployment.yaml
vim manifests/kube-dns-deployment.yaml (add label)
kubectl -n kube-system alpha diff -f manifests/kube-dns-deployment.yaml
```